core/src/main/java/org/springframework/security/core/SupplierAuthentication.java

```diff
@@ -0,0 +1,147 @@
+/*
+ * Copyright 2002-2022 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.springframework.security.core;
+
+import java.util.Collection;
+import java.util.function.Supplier;
+
+import org.springframework.util.Assert;
+
+/**
+ * An {@link Authentication} that lazily initializes another {@link Authentication}.
+ *
+ * @author Evgeniy Cheban
+ * @since 5.7
+ */
+public final class SupplierAuthentication implements Authentication {
```

Comment:
Cool solution. Instead of doing this, I'd prefer to see if we can keep contracts explicit to it's clear that a given API is deferring the lookup.

If this is ultimately needed to make `SecurityExpressionRoot` work, then maybe it can be inlined to that class.

Reply:
I agree. It would be much easier if we can change `SecurityExpressionRoot.authentication` to `private Supplier<Authentication>`, for those users who use `this.authentication` shouldn't be too difficult to change their code to `getAuthentication()`.

If we'll go with `private Supplier<Authentication>` approach it might be reasonable to introduce a `Supplier` that memorises the value:
```
private static final class AuthenticationSupplier implements Supplier<Authentication> {
    private final Supplier<Authentication> delegate;
    private volatile Authentication value;
    
    @Override
    public Authentication get() {
	if (this.value == null) {
	    synchronized (this.delegate) {
		if (this.value == null) {
		    this.value = this.delegate.get();;
	        }
            }
	}
	return this.value;
    }
}
```
Therefore for the `expression` like `hasRole('USER') or hasRole('ADMIN')` we only obtain an `Authentication` object once instead of calling `Supplier#get` for every `hasRole` in the `expression`.